### PR TITLE
feat(web/applications): add applicant address step (boas-152)

### PR DESCRIPTION
**Pull request by lutamKC**
## Describe your changes

Add step on applicant form to add address.
1. find list of address by postcode using an extenral service
2. select on the addresses from a dropdown
3. if none of the addresses in the list is correct, enter address manually

also includes:
- refactor of the nunjuck url filter

## Type of change 🧩

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (please explain in the description section above)

## Checklist before requesting a review

- [ ] I have performed a self-review of my own code
- [ ] I have double checked this work does not include any hardcoded secrets or passwords
- [ ] I have made corresponding changes to the documentation
- [ ] I have provided details on how I have tested my code
- [ ] I have referenced the ticket number above
- [ ] I have provided a description of the ticket
- [ ] I have included unit tests to cover any testable code changes

### CODE REVIEW

File: apps/web/src/server/applications/create/applicant/applications-create-applicant.types.d.ts [hunk 1]
```diff
@@ -34,6 +34,26 @@ export type ApplicationsCreateApplicantFullNameBody = {
...
+	| 'selectAddress'
+	| 'manualAddress';
+export type ApplicationsCreateApplicantAddressProps = {
+	formStage: ApplicationCreateApplicantAddressStage;
+	errors?: Record<string, { msg: string }> | ValidationErrors;
+	postcode: string;
+	addressList?: ApplicationAddress[];
+};
+
+export type ApplicationsCreateApplicantAddressBody = {
```
Comment (lutamKC) [inline apps/web/src/server/applications/create/applicant/applications-create-applicant.types.d.ts:48]: i'm not proud of this, but this page is actually three pages so the structure of `body` can be quite diffrent

File: apps/web/src/server/appeals/inspector/__tests__/__snapshots__/inspector.test.js.snap [hunk 2]
```diff
@@ -383,7 +383,7 @@ exports[`inspector GET /appeals/:appealId should render the appeal details 1`] =
                 </div>
             </div>
         </div>
-        
+
```
Comment (lutamKC) [inline apps/web/src/server/appeals/inspector/__tests__/__snapshots__/inspector.test.js.snap:386]: ?

**Comment by francescobrenci-AR** [comment]
can you add the API_KEY to `env.test` and `env.production` please

File: apps/web/src/server/applications/create/applicant/applications-create-applicant.controller.js [hunk 3]
```diff
@@ -135,27 +142,93 @@ export async function updateApplicationsCreateApplicantEmail(
...
+				const { errors, addressList } = await findAddressListByPostcode(postcode, {
+					maxResults: 50,
+					minMatch: 0.9
+				});
+				const newFormStage = errors ? 'searchPostcode' : 'selectAddress';
+
+				return { errors, addressList, formStage: newFormStage };
+			}
+			case 'selectAddress': {
+				const { errors: serviceErrors, addressList } = await findAddressListByPostcode(postcode, {
```
Comment (francescobrenci-AR) [inline apps/web/src/server/applications/create/applicant/applications-create-applicant.controller.js:184]: do you need to call `findAddressListByPostcode` also when you show the results?

Comment (lutamKC) [reply apps/web/src/server/applications/create/applicant/applications-create-applicant.controller.js:184]: yes, we need to show them the addresses, to compare the selected apiReference with the list and finally to actually send the data to the db.
We could save them in the session in the previous step but I'm trying to use the session as little as possible and then I was thinking that we'll need to probably make a request anyway when we resume it

**Comment by francescobrenci-AR** [comment]
Now that the API KEY is defined in the infrastructure here https://github.com/Planning-Inspectorate/infrastructure-environments/pull/124, please add it to our env files as discussed (OS_PLACES_API_KEY = os_places_api_key) and use this var.

We then need a env.local (not committed to GitHub) with the hard coded key for us to work locally

File: apps/web/src/server/applications/applications.types.d.ts [hunk 4]
```diff
@@ -26,6 +26,19 @@ export interface Application {
...
+	id: number;
+	address?: ApplicationsAddress;
+}
+// TODO: unify with appeals address
+export interface ApplicationsAddress {
+	postcode: string;
+	addressLine1: string;
+	addressLine2?: string;
+	town: string;
 }
```
Comment (mayya-bond-kc) [inline apps/web/src/server/applications/applications.types.d.ts:42]: no county?

**Review by mayya-bond-kc — APPROVED** [review]

**Review by francescobrenci-AR — APPROVED** [review]

File: apps/web/src/server/applications/applications.types.d.ts (hunk 4 above)
Comment (francescobrenci-AR) [reply apps/web/src/server/applications/applications.types.d.ts:42]: apparently that's not needed

**Pull request merged**